fork_choice/confirmation-rule.md

```diff
@@ -0,0 +1,673 @@
+    checkpoint = Checkpoint(root=checkpoint_root, epoch=block_epoch)
+
+    if checkpoint not in store.checkpoint_states:
+        return Gwei(0)
+
+    checkpoint_state = store.checkpoint_states[checkpoint]
+
+    leave_roots = get_leaf_block_roots(store, block_root)
+
+    active_validator_indices = get_active_validator_indices(checkpoint_state, block_epoch)
+    participating_indices = set().union(*[
+        get_epoch_participating_indices(
+            store.block_states[root], 
+            active_validator_indices, 
+            block_epoch == current_epoch
+        )
+        for root in leave_roots
+    ])
+
+    return get_total_balance(checkpoint_state, participating_indices)
+```
+
+
+
+#### `is_ffg_confirmed_current_epoch`
+
+```python
+def is_ffg_confirmed_current_epoch(
+    total_active_balance: int,
+    ffg_support_for_checkpoint: int,
+    remaining_ffg_weight: int,
+    confirmation_byzantine_threshold: int,
+    confirmation_slashing_threshold: int,   
+) -> bool:
+    """
+    Returns whether the branch will justify it's current epoch checkpoint at the end of this epoch
+    """
+
+    max_adversarial_ffg_support_for_checkpoint = int(
+        min(
+            (total_active_balance * confirmation_byzantine_threshold - 1) // 100 + 1,
+            confirmation_slashing_threshold,
+            ffg_support_for_checkpoint
+        )
+    )
+
+    """
+    Returns whether the following condition is true using only integer arithmetic
+    2 / 3 * total_active_balance <= (
+        ffg_support_for_checkpoint - max_adversarial_ffg_support_for_checkpoint +
+        (1 - confirmation_byzantine_threshold / 100) * remaining_ffg_weight
+    )
+    """
+
+    return (
+        200 * total_active_balance <=
+        ffg_support_for_checkpoint * 300 + (300 - 3 * confirmation_byzantine_threshold) *
+        remaining_ffg_weight - max_adversarial_ffg_support_for_checkpoint * 300
+    )
+```
+
+#### `is_ffg_confirmed_previous_epoch`
+
+```python
+def is_ffg_confirmed_previous_epoch(
+    total_active_balance: int,
+    ffg_support_for_checkpoint: int,
+    confirmation_byzantine_threshold: int,
+    confirmation_slashing_threshold: int,    
+) -> bool:
+    """
+    Returns whether the `block_root`'s checkpoint is justified
+    """
+
+    max_adversarial_ffg_support_for_checkpoint = int(
+        min(
+            (total_active_balance * confirmation_byzantine_threshold - 1) // 100 + 1,
+            confirmation_slashing_threshold,
+            ffg_support_for_checkpoint
+        )
+    )    
+
+    return 2 * total_active_balance <= (ffg_support_for_checkpoint - max_adversarial_ffg_support_for_checkpoint) * 3
+```
+
+#### `is_ffg_confirmed`
+
+```python
+def is_ffg_confirmed(
+    store: Store,
+    confirmation_byzantine_threshold: int,
+    confirmation_slashing_threshold: int,
+    block_root: Root,
+) -> bool:
+    """
+    Returns whether the `block_root`'s checkpoint is justified
```

Comment:
this is the same exact docstring as `is_ffg_confirmed_previous_epoch`. Let's make sure to clarify

Also `is_ffg_confirmed_[previous/current]_epoch` could use the same function and signature if  it was conditioned on `remaining_ffg_weight == 0`

Reply:
Changed the doc.